### PR TITLE
[FLINK-5968] Add documentation for WindowedStream.aggregate()

### PR DESCRIPTION
**Pull request by aljoscha**

### CODE REVIEW

**Comment by aljoscha** [comment]
R: @twalthr you might be interested 😃

File: docs/dev/stream/operators/windows.md [hunk 1]
```diff
@@ -721,6 +808,111 @@ input
...
+}
+
+private static class MyProcessWindowFunction
+    implements ProcessWindowFunction<Double, Tuple2<String, Double>, String, TimeWindow> {
+
+  public void apply(String key,
+                    Context context,
+                    Iterable<Double> averages,
+                    Collector<Tuple2<String, Double>> out) {
+      Double average = averags.iterator().next();
```
Comment (alpinegizmo) [inline docs/dev/stream/operators/windows.md]: should be averages, not averags

Comment (aljoscha) [reply docs/dev/stream/operators/windows.md]: Thanks! fixing

**Comment by aljoscha** [comment]
CC: @fhueske You wanted me to cc you for the Table API changes.

File: flink-libraries/flink-table/src/main/scala/org/apache/flink/table/runtime/aggregate/GeneratedAggregations.scala [hunk 2]
```diff
@@ -67,7 +67,7 @@ abstract class GeneratedAggregations extends Function {
     *                     aggregated results
     * @param input        input values bundled in a row
```
Comment (fhueske) [inline flink-libraries/flink-table/src/main/scala/org/apache/flink/table/runtime/aggregate/GeneratedAggregations.scala]: Please add `@returns` documentation

Comment (fhueske) [reply flink-libraries/flink-table/src/main/scala/org/apache/flink/table/runtime/aggregate/GeneratedAggregations.scala]: Please change `retract()` accordingly to keep the interfaces consistent.

Alternatively, you can also revert the changes on `GeneratedAggregations` which is only an internal interface. The Table API `org.apache.flink.table.functions.AggregateFunction` does not support immutable accumulators, so this change has no effect until the public interface is changed. If you revert, we only need to touch `AggregateAggFunction` and not the code generator.

File: docs/dev/stream/operators/windows.md [hunk 3]
```diff
@@ -427,6 +427,93 @@ input
...
+input type (`IN`), accumulator type (`ACC`), and an output type (`OUT`). The input type is the type
+of elements in the input stream and the `AggregateFunction` has a method for adding one input
+element to an accumulator. The interface also has methods for creating an initial accumulator,
+for merging two accumulators into one accumulator and for extracting an output (of type `OUT`) from
+an accumulator. We will see how this works in the example below.
+
+Same as with `ReduceFunction`, Flink will incrementally aggregate input elements of a window as they
+arrive.
+
+A `AggregateFunction` can be defined and used like this:
```
Comment (bowenli86) [inline docs/dev/stream/operators/windows.md]: An

File: docs/dev/stream/operators/windows.md [hunk 4]
```diff
@@ -427,6 +427,93 @@ input
...
+private static class AverageAggregate
+    implements AggregateFunction<Tuple2<String, Long>, Tuple2<Long, Long>, Double> {
+  @Override
+  public Tuple2<Long, Long> createAccumulator() {
+    return new Tuple2<>(0L, 0L);
+  }
+
+  @Override
+  public Tuple2<Long, Long> add(
+    Tuple2<String, Long> value, Tuple2<Long, Long> accumulator) {
```
Comment (bowenli86) [inline docs/dev/stream/operators/windows.md]: shall we keep the method signature in a single line? They can fit into one line, and I usually find it difficult to read method signature in two lines

File: docs/dev/stream/operators/windows.md [hunk 5]
```diff
@@ -427,6 +427,93 @@ input
...
+  }
+
+  @Override
+  public Double getResult(Tuple2<Long, Long> accumulator) {
+    return accumulator.f0 / accumulator.f1;
+  }
+
+  @Override
+  public Tuple2<Long, Long> merge(
+    Tuple2<Long, Long> a, Tuple2<Long, Long> b) {
```
Comment (bowenli86) [inline docs/dev/stream/operators/windows.md]: ditto

File: docs/dev/stream/operators/windows.md [hunk 6]
```diff
@@ -721,6 +808,111 @@ input
...
+private static class AverageAggregate
+    implements AggregateFunction<Tuple2<String, Long>, Tuple2<Long, Long>, Double> {
+  @Override
+  public Tuple2<Long, Long> createAccumulator() {
+    return new Tuple2<>(0L, 0L);
+  }
+
+  @Override
+  public Tuple2<Long, Long> add(
+    Tuple2<String, Long> value, Tuple2<Long, Long> accumulator) {
```
Comment (bowenli86) [inline docs/dev/stream/operators/windows.md]: ditto

File: docs/dev/stream/operators/windows.md [hunk 7]
```diff
@@ -721,6 +808,111 @@ input
...
+  }
+
+  @Override
+  public Double getResult(Tuple2<Long, Long> accumulator) {
+    return accumulator.f0 / accumulator.f1;
+  }
+
+  @Override
+  public Tuple2<Long, Long> merge(
+    Tuple2<Long, Long> a, Tuple2<Long, Long> b) {
```
Comment (bowenli86) [inline docs/dev/stream/operators/windows.md]: ditto

**Comment by bowenli86** [comment]
I just saw this PR, and sorry for submitting my feedbacks late...

**Comment by aljoscha** [comment]
Thanks, @bowenli86! Your recommendations are good and I'll push a hot fix commit with the fixes.

**Pull request closed without merging**